# DeepChem Step-by-Step Tutorial

In this tutorial series, you'll learn how to use DeepChem to solve interesting
and challenging problems in the life sciences. The tutorial acts as a introduction
to DeepChem as well as application of DeepChem to a wide array of problems
across domains like molecular machine learning,
quantum chemistry, bioinformatics and material science. This tutorial series is
continually updated with new DeepChem features and models as implemented and is
designed to be accessible to beginners.

## Why do the DeepChem Tutorial?

**1) Career Advancement:** Applying AI in the life sciences is a booming
industry at present. There are a host of newly funded startups and initiatives
at large pharmaceutical and biotech companies centered around AI. Learning and
mastering DeepChem will bring you to the forefront of this field and will
prepare you to enter a career in this field.

**2) Humanitarian Considerations:** Disease is the oldest cause of human
suffering. From the dawn of human civilization, humans have suffered from pathogens,
cancers, and neurological conditions. One of the greatest achievements of
the last few centuries has been the development of effective treatments for
many diseases. By mastering the skills in this tutorial, you will be able to
stand on the shoulders of the giants of the past to help develop new
medicine.

**3) Lowering the Cost of Medicine:** The art of developing new medicine is
currently an elite skill that can only be practiced by a small core of expert
practitioners. By enabling the growth of open source tools for drug discovery,
you can help democratize these skills and open up drug discovery to more
competition. Increased competition can help drive down the cost of medicine.

The tutorial is organized as follows:

### Introduction to DeepChem
The following tutorials covers the core features of DeepChem. Covering these
tutorials will help you in getting started with DeepChem for machine learning. The later
tutorials discuss about using DeepChem for specific applications.

* [1 The Basic Tools of the Deep Life Sciences](The_Basic_Tools_of_the_Deep_Life_Sciences.ipynb)
* [2 Working With Datasets](Working_With_Datasets.ipynb)
* [3 An Introduction to MoleculeNet](An_Introduction_To_MoleculeNet.ipynb)
* [4 Molecular Fingerprints](Molecular_Fingerprints.ipynb)
* [5 Creating Models with Tensorflow and Pytorch](Creating_Models_with_TensorFlow_and_PyTorch.ipynb)
* [6 Introduction to Graph Convolutions](Introduction_to_Graph_Convolutions.ipynb)
* [7 Going Deeper on Molecular Featurizations](Going_Deeper_on_Molecular_Featurizations.ipynb)
* [8 Working With Splitters](Working_With_Splitters.ipynb)
* [9 Advanced Model Training](Advanced_Model_Training.ipynb)
* [10 Creating a high fidelity model from experimental data](Creating_a_high_fidelity_model_from_experimental_data.ipynb)
* [11 Putting Multitask Learning to Work](Putting_Multitask_Learning_to_Work.ipynb)
* [12 Modeling Protein Ligand Interactions](Modeling_Protein_Ligand_Interactions.ipynb)
* [13 Modeling Protein Ligand Interactions with Atomic Convolutions](Modeling_Protein_Ligand_Interactions_With_Atomic_Convolutions.ipynb)
* [14 Conditional Generative Adversarial Networks](Conditional_Generative_Adversarial_Networks.ipynb)
* [15 Training a Generative Adversarial Network on MNIST](Training_a_Generative_Adversarial_Network_on_MNIST.ipynb)
* [16 Distributed Multi-GPU Training of DeepChem Models with LitMatter](https://github.com/ncfrey/litmatter/blob/main/LitDeepChem.ipynb)
* [17 Advanced model training using Hyperopt](Advanced_model_training_using_hyperopt.ipynb)
* [18 Introduction to Gaussian Processes](Introduction_to_Gaussian_Processes.ipynb)
* [19 Pytorch-Lightning Integration for DeepChem Models](PytorchLightning_Integration.ipynb)
* [20 Compiling DeepChem PyTorch Models](Compiling_DeepChem_Torch_Models.ipynb)

### Molecular Machine Learning
* [1 Molecular Fingerprints](Molecular_Fingerprints.ipynb)
* [2 Going Deeper on Molecular Featurizations](Going_Deeper_on_Molecular_Featurizations.ipynb)
* [3 Learning Unsupervised Embeddings for Molecules](Learning_Unsupervised_Embeddings_for_Molecules.ipynb)
* [4 Synthetic Feasibility Scoring](Synthetic_Feasibility_Scoring.ipynb)
* [5 Atomic Contributions for Molecules](Atomic_Contributions_for_Molecules.ipynb)
* [6 Interactive Model Evaluation with Trident Chemwidgets](Interactive_Model_Evaluation_with_Trident_Chemwidgets.ipynb)
* [7 Transfer Learning with ChemBERTa Transformers](Transfer_Learning_With_ChemBERTa_Transformers.ipynb)
* [8 Training a Normalizing Flow on QM9](Training_a_Normalizing_Flow_on_QM9.ipynb)
* [9 Large Scale Chemical Screens](Large_Scale_Chemical_Screens.ipynb)
* [10 Introduction to Molecular Attention Transformer](Introduction_to_Molecular_Attention_Transformer.ipynb)
* [11 Generating Molecules with MolGAN](Generating_molecules_with_MolGAN.ipynb)
* [12 Introduction to GROVER](Introduction_to_GROVER.ipynb)

### Modeling Proteins
* [1 Protein Deep Learning](Protein_Deep_Learning.ipynb)
* [2 Antibody Design via Protein Language Models and Directed Evolution](DeepChem_AntibodyTutorial_Simplified.ipynb)
* [3 Protein Structure Prediction with ESMFold](Protein_Structure_Prediction_with_ESMFold.ipynb)
* [4 Introduction to ProtBERT](Introduction_to_ProtBERT.ipynb)
* [5 Protein Language Models (Intuition)](ProteinLM_Tutorial0.ipynb)

### Protein Ligand Modeling
* [1 Modeling Protein Ligand Interactions](Modeling_Protein_Ligand_Interactions.ipynb)
* [2 Modeling Protein Ligand Interactions with Atomic Convolutions](Modeling_Protein_Ligand_Interactions_With_Atomic_Convolutions.ipynb)
* [3 Applications of DeepChem with Alphafold](DeepChemXAlphafold.ipynb)

### Quantum Chemistry
* [1 Exploring Quantum Chemistry with GDB1k](Exploring_Quantum_Chemistry_with_GDB1k.ipynb)
* [2 Finding ground state energy of molecules with DeepQMC](DeepQMC_tutorial.ipynb)
* [3 Training an exchange correlation functional using DeepChem](Training_an_Exchange_Correlation_Functional_using_Deepchem.ipynb)
* [4 Finding ground state energy of molecules with FermiNet(DeepChem Implementation)](FermiNet_DeepChem_tutorial.ipynb)

### Bioinformatics
* [1 Introduction to BioInformatics](Introduction_to_Bioinformatics.ipynb)
* [2 Multisequence Alignments](Multisequence_Alignments.ipynb)
* [3 Scanpy](Scanpy.ipynb)
* [4 Deep probabilistic analysis of omics data](Deep_probabilistic_analysis_of_single-cell_omics_data.ipynb)
* [5 Automated Cell Counting](Cell_Counting_Tutorial.ipynb)

### Material Science
* [1 Introduction to Material Science](Introduction_To_Material_Science.ipynb)

### Machine Learning Methods
* [1 Using Reinforcement Learning to Play Pong](Using_Reinforcement_Learning_to_Play_Pong.ipynb)
* [2 Introduction to Model Interpretability](Introduction_to_Model_Interpretability.ipynb)
* [3 Uncertainty in Deep Learning](Uncertainty_In_Deep_Learning.ipynb)

### Deep Differential Equations
* [1 Physics Informed Neural Networks (Burgers Equation)](Physics_Informed_Neural_Networks.ipynb)
* [2 Introducing_JaxModel_and_PINNModel](Introducing_JaxModel_and_PINNModel.ipynb)
* [3 About_nODE_Using_Torchdiffeq_in_Deepchem](About_nODE_Using_Torchdiffeq_in_Deepchem.ipynb)

### Equivariance
* [1 Introduction to Equivariance](Introduction_to_Equivariance.ipynb)

### Olfaction
* [1 Predict Multi-Label Odor Descriptors using OpenPOM](Predict_Multi_Label_Odor_Descriptors_using_OpenPOM.ipynb)

### PROTACs
* [1 Introduction to PROTACs](Introduction_to_PROTACs.ipynb)

### Polymers
* [1 An Introduction to the Polymers and Their Representation](An_Introduction_to_the_Polymers_and_Their_Representation.ipynb)
* [2 Utilizing Deepchem Molecular Models for Crystalization Tendency Regression of Polymers](crystallization_tendency_regression.ipynb)
* [3 Introduction to Polymer SMILES](Introduction_to_Polymer_SMILES.ipynb)
* [4 Understanding Weighted Directed Graphs for Polymer Implementations](Understanding_Weighted_Directed_Graphs_for_Polymer_Implimentations.ipynb)
* [5 PolyBERT: Introduction to Chemical Language Models for Fingerprint Generation](PolyBERT___Introduction_to_Chemical_Language_Models_for_Fingerprint_Generation.ipynb)